File: processor/src/geotrace/parking_tracker.py
"""Independent free-space tracker for the last manoeuvre before parking.

Unlike :mod:`geotrace.road_tracker`, this state is ENU and its signed velocity
is never clipped.  It consumes only recorded IMU/GPS plus a road prior; a
reference track is not accepted anywhere in this module.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Optional, Sequence

import numpy as np

from geotrace.config import Config
from geotrace.coordinates import LocalFrame, course_to_heading, wrap_angle
from geotrace.models import LocationSample
from geotrace.motion_model import ImuControl, longitudinal_acceleration


@dataclass
class ParkingResult:
    status: str
    position: tuple[float, float]
    covariance: np.ndarray
    confidence: float
    polygon_radius_m: float
    times: list[float] = field(default_factory=list)
    trajectory: list[tuple[float, float]] = field(default_factory=list)
    terminal_cluster_count: int = 0
    rejected_fixes: int = 0
    has_reverse_motion: bool = False
    reason: str = ""
    max_distance_from_road_m: Optional[float] = None

    def to_json(self, frame: LocalFrame) -> dict:
        lat, lon = frame.to_geo(*self.position)
        return {
            "status": self.status,
            "position": {"latitude": lat, "longitude": lon},
            "covariance_m2": self.covariance.round(3).tolist(),
            "confidence": round(self.confidence, 3),
            "polygon_radius_m": round(self.polygon_radius_m, 2),
            "terminal_cluster_fixes": self.terminal_cluster_count,
            "rejected_fixes": self.rejected_fixes,
            "has_reverse_motion": self.has_reverse_motion,
            "reason": self.reason,
            "max_distance_from_road_m": self.max_distance_from_road_m,
        }


class ParkingTracker:
    """Six-state, free-space terminal tracker: ``x,y,u,theta,b_a,b_w``."""

    def __init__(self, cfg: Config, frame: LocalFrame) -> None:
        self.cfg = cfg
        self.frame = frame
        self.state = np.zeros(6, dtype=float)
        self.P = np.diag([25.0, 25.0, 4.0, 0.8, 0.3, 0.03])

    def run(
        self,
        controls: Sequence[ImuControl],
        fixes: Sequence[LocationSample],
        ended_at: float,
        road_prior_xy: Sequence[float],
        road_heading: float,
        road_speed: float,
        calibration_present: bool,
        road_distance: Optional[callable] = None,
    ) -> ParkingResult:
        cfg = self.cfg.parking_tracker
        start = ended_at - cfg.window_s
        controls = [c for c in controls if start <= c.t <= ended_at]
        fixes = [f for f in fixes if start <= f.monotonic_time <= ended_at and f.is_usable]
        self.state[:] = [road_prior_xy[0], road_prior_xy[1], road_speed, road_heading, 0.0, 0.0]
        trajectory: list[tuple[float, float]] = []
        times: list[float] = []
        accepted: list[tuple[LocationSample, np.ndarray]] = []
        rejected = 0
        cursor = 0
        reverse = False
        for control in controls:
            self._predict(control)
            reverse = reverse or self.state[2] < -0.25
            while cursor < len(fixes) and fixes[cursor].monotonic_time <= control.t:
                fix = fixes[cursor]
                cursor += 1
                xy = np.asarray(self.frame.to_local(fix.latitude, fix.longitude), dtype=float)
                if not self._reachable(fix, xy, accepted, road_prior_xy, start):
                    rejected += 1
                    continue
                self._gps_update(fix, xy)
                accepted.append((fix, xy))
            trajectory.append((float(self.state[0]), float(self.state[1])))
            times.append(control.t)

        # No prediction after `ended_at`: signed velocity is a terminal stop
        # constraint, never an extra extrapolation after recording stopped.
        self.state[2] = 0.0
        terminal = self._terminal_cluster(accepted, ended_at)
        if terminal:
            points, weights = terminal
            centre = np.average(points, axis=0, weights=weights)
            scatter = float(np.sqrt(np.average(np.sum((points - centre) ** 2, axis=1), weights=weights)))
            radius = max(4.0, scatter * 2.0, math.sqrt(1.0 / weights.sum()) * 2.0)
            confidence = min(0.95, 0.55 + 0.08 * len(points) + (0.08 if calibration_present else 0.0))
            status = "CONFIDENT" if confidence >= 0.78 else "PROBABLE"
            reason = "robust low-speed terminal GPS cluster"
            self.state[:2] = centre
        elif accepted:
            radius, confidence, status = cfg.uncertain_radius_m, 0.32, "UNCERTAIN"
            reason = "no mutually consistent terminal GPS cluster"
        else:
            radius, confidence, status = cfg.uncertain_radius_m, 0.12, "INSUFFICIENT_DATA"
            reason = "no usable GPS in terminal window"
        if not calibration_present:
            confidence *= 0.65
            if status == "CONFIDENT":
                status = "PROBABLE"
            reason += "; mount calibration unavailable"
        max_road = None
        if road_distance and trajectory:
            max_road = float(max(road_distance(p) for p in trajectory))
        return ParkingResult(
            status=status, position=(float(self.state[0]), float(self.state[1])),
            covariance=np.diag([radius**2, radius**2]), confidence=confidence,
            polygon_radius_m=radius, times=times, trajectory=trajectory,
            terminal_cluster_count=(len(terminal[0]) if terminal else 0), rejected_fixes=rejected,
            has_reverse_motion=bool(reverse), reason=reason, max_distance_from_road_m=max_road,
        )

    def _predict(self, control: ImuControl) -> None:
        if control.dt <= 0 or control.gap_exceeded:
            return
        x, y, u, theta, ba, bw = self.state
        theta = float(wrap_angle(theta + (control.yaw_rate - bw) * control.dt))
        a = longitudinal_acceleration(control.a_world, theta) - ba
        ds = u * control.dt + 0.5 * a * control.dt * control.dt
        self.state[0] = x + ds * math.cos(theta)
        self.state[1] = y + ds * math.sin(theta)
        self.state[2] = u + a * control.dt  # signed: reverse is valid
        self.state[3] = theta
        self.P[:2, :2] += np.eye(2) * (self.cfg.parking_tracker.process_position_sigma_m * control.dt) ** 2

    def _reachable(self, fix, xy, accepted, road_prior_xy, start) -> bool:
        cfg = self.cfg.parking_tracker
        if fix.horizontal_accuracy is None or fix.horizontal_accuracy <= 0 or fix.horizontal_accuracy > cfg.max_accuracy_m:
            return False
        if accepted:
            prev, prev_xy = accepted[-1]
            dt = max(0.0, fix.monotonic_time - prev.monotonic_time)
            maximum = self.cfg.motion.max_speed_ms * dt + cfg.physical_margin_m
            return math.dist(xy, prev_xy) <= maximum
        dt = max(0.0, fix.monotonic_time - start)
        maximum = self.cfg.motion.max_speed_ms * dt + cfg.physical_margin_m
        return math.dist(xy, road_prior_xy) <= maximum

    def _gps_update(self, fix: LocationSample, xy: np.ndarray) -> None:
        sigma = max(float(fix.horizontal_accuracy or 8.0), self.cfg.gps.min_accuracy_sigma_m)
        gain = float(np.clip(30.0 / (30.0 + sigma * sigma), 0.08, 0.65))
        previous = self.state[:2].copy()
        self.state[:2] += gain * (xy - self.state[:2])
        if fix.has_valid_course and fix.has_valid_speed and float(fix.speed or 0.0) >= self.cfg.parking_tracker.low_speed_ms:
            self.state[3] = course_to_heading(float(fix.course or 0.0))
        if fix.has_valid_speed:
            observed_speed = float(fix.speed or 0.0)
            displacement = xy - previous
            direction = np.array([math.cos(self.state[3]), math.sin(self.state[3])])
            sign = -1.0 if float(np.dot(displacement, direction)) < 0 else 1.0
            self.state[2] = sign * observed_speed

    def _terminal_cluster(self, accepted, ended_at):
        cfg = self.cfg.parking_tracker
        tail = [(f, xy) for f, xy in accepted if f.monotonic_time >= ended_at - cfg.terminal_cluster_window_s
                and f.has_valid_speed and float(f.speed or 0.0) <= cfg.low_speed_ms]
        if len(tail) < cfg.min_cluster_fixes:
            return None
        points = np.asarray([xy for _, xy in tail])
        median = np.median(points, axis=0)
        kept = [(f, xy) for f, xy in tail if math.dist(xy, median) <= 35.0]
        if len(kept) < cfg.min_cluster_fixes:
            return None
        weights = np.asarray([1.0 / max(float(f.horizontal_accuracy or 5.0), 5.0) ** 2 for f, _ in kept])
        return np.asarray([xy for _, xy in kept]), weights

File: ios/LiveTracking/Sources/LiveTracking/ParkingTracker.swift
import Foundation

/// Independent free-space tracker for the last manoeuvre before parking.
///
/// A port of `geotrace.parking_tracker.ParkingTracker`. Its state is
///
///     X = [x, y, u, theta, b_a, b_omega]
///
/// in the local ENU frame, where `u` is a **signed** speed. Nothing clamps it
/// with `max(0, u)`: a car reversing into a space is moving backwards, and a
/// tracker that cannot represent that will place it several metres past where it
/// actually stopped.
///
/// Two things separate it from the road tracker and are deliberate:
/// it may leave the OSM graph entirely (courtyards, car parks, forecourts), and
/// it never writes anything back into the road belief.
///
/// The evaluation is a pure function of the rolling window, so the live runtime
/// can simply re-run it as the window slides. That also means it is exactly the
/// batch algorithm the Python baseline runs, with no incremental approximation
/// to drift apart from it.
public struct ParkingTracker {
    private let config: LiveTrackingConfig
    private let frame: LocalFrame

    public init(config: LiveTrackingConfig, frame: LocalFrame) {
        self.config = config
        self.frame = frame
    }

    /// Six-state vector `[x, y, u, theta, b_a, b_omega]`.
    public struct State {
        public var x = 0.0, y = 0.0
        public var u = 0.0
        public var theta = 0.0
        public var accelBias = 0.0
        public var gyroBias = 0.0
        public var positionVariance = 25.0

        public init() {}

        var point: Point { Point(x: x, y: y) }
    }

    /// Test seams: the transition and the measurement update are the two places
    /// signed velocity is decided, and they are worth asserting on directly
    /// rather than only through a whole manoeuvre.
    public func testPredict(_ state: inout State, control: IMUControl) {
        predict(&state, control: control)
    }

    public func testGPSUpdate(_ state: inout State, fix: TrackerLocation, point: Point) {
        gpsUpdate(&state, fix: fix, point: point)
    }

    public struct Inputs {
        public var controls: [IMUControl]
        public var fixes: [TrackerLocation]
        public var endedAt: Double
        public var roadPrior: Point
        public var roadHeading: Double
        public var roadSpeed: Double
        public var calibrationPresent: Bool
        public var roadDistance: ((Point) -> Double)?

        public init(
            controls: [IMUControl], fixes: [TrackerLocation], endedAt: Double,
            roadPrior: Point, roadHeading: Double, roadSpeed: Double,
            calibrationPresent: Bool, roadDistance: ((Point) -> Double)? = nil
        ) {
            self.controls = controls
            self.fixes = fixes
            self.endedAt = endedAt
            self.roadPrior = roadPrior
            self.roadHeading = roadHeading
            self.roadSpeed = roadSpeed
            self.calibrationPresent = calibrationPresent
            self.roadDistance = roadDistance
        }
    }

    public func evaluate(_ inputs: Inputs) -> ParkingTrackingResult {
        let cfg = config.parkingTracker
        let start = inputs.endedAt - cfg.windowSeconds
        let controls = inputs.controls.filter { $0.t >= start && $0.t <= inputs.endedAt }
        let fixes = inputs.fixes
            .filter { $0.monotonicTime >= start && $0.monotonicTime <= inputs.endedAt && $0.isUsable }
            .sorted { $0.monotonicTime < $1.monotonicTime }

        var state = State()
        state.x = inputs.roadPrior.x
        state.y = inputs.roadPrior.y
        state.u = inputs.roadSpeed
        state.theta = inputs.roadHeading

        var trajectory: [Point] = []
        var accepted: [(fix: TrackerLocation, point: Point)] = []
        var rejected = 0
        var cursor = 0
        var hasReverse = false

        for control in controls {
            predict(&state, control: control)
            if state.u < cfg.reverseSpeedThresholdMS { hasReverse = true }
            while cursor < fixes.count && fixes[cursor].monotonicTime <= control.t {
                let fix = fixes[cursor]
                cursor += 1
                let point = frame.toLocal(latitude: fix.latitude, longitude: fix.longitude)
                guard reachable(
                    fix: fix, point: point, accepted: accepted,
                    roadPrior: inputs.roadPrior, start: start
                ) else {
                    rejected += 1
                    continue
                }
                gpsUpdate(&state, fix: fix, point: point)
                accepted.append((fix, point))
            }
            trajectory.append(state.point)
        }

        // No prediction after `endedAt`: the end of recording is a terminal
        // stop constraint, never an excuse to extrapolate further.
        state.u = 0

        var radius: Double
        var confidence: Double
        var status: ParkingStatus
        var reason: String
        var clusterCount = 0

        if let cluster = terminalCluster(accepted: accepted, endedAt: inputs.endedAt) {
            let weightSum = cluster.weights.reduce(0, +)
            var centre = Point.zero
            for (point, weight) in zip(cluster.points, cluster.weights) {
                centre = Point(x: centre.x + point.x * weight, y: centre.y + point.y * weight)
            }
            centre = Point(x: centre.x / weightSum, y: centre.y / weightSum)
            var scatterSum = 0.0
            for (point, weight) in zip(cluster.points, cluster.weights) {
                let dx = point.x - centre.x, dy = point.y - centre.y
                scatterSum += (dx * dx + dy * dy) * weight
            }
            let scatter = (scatterSum / weightSum).squareRoot()
            radius = Swift.max(4.0, scatter * 2.0, (1.0 / weightSum).squareRoot() * 2.0)
            confidence = Swift.min(
                0.95,
                0.55 + 0.08 * Double(cluster.points.count)
                    + (inputs.calibrationPresent ? 0.08 : 0.0)
            )
            status = confidence >= 0.78 ? .confident : .probable
            reason = "robust low-speed terminal GPS cluster"
            clusterCount = cluster.points.count
            state.x = centre.x
            state.y = centre.y
        } else if !accepted.isEmpty {
            radius = cfg.uncertainRadiusM
            confidence = 0.32
            status = .uncertain
            reason = "no mutually consistent terminal GPS cluster"
        } else {
            radius = cfg.uncertainRadiusM
            confidence = 0.12
            status = .insufficientData
            reason = "no usable GPS in terminal window"
        }

        if !inputs.calibrationPresent {
            confidence *= 0.65
            if status == .confident { status = .probable }
            reason += "; mount calibration unavailable"
        }

        var maxRoadDistance: Double?
        if let roadDistance = inputs.roadDistance, !trajectory.isEmpty {
            maxRoadDistance = trajectory.map(roadDistance).max()
        }

        // With genuinely nothing to go on there is no honest position to
        // report. Drawing the road endpoint here instead would be a lie: the
        // road tracker is pinned to the carriageway and cannot represent a car
        // parked in a courtyard.
        let hasPosition = status != .insufficientData
        let position = hasPosition ? frame.toGeo(state.point) : nil

        let lastFixGap = accepted.last.map { inputs.endedAt - $0.fix.monotonicTime }

        return ParkingTrackingResult(
            status: status,
            position: position,
            positionLocal: hasPosition ? state.point : nil,
            confidence: confidence,
            polygonRadiusM: radius,
            polygon: hasPosition ? circle(around: state.point, radius: radius) : [],
            terminalClusterFixes: clusterCount,
            rejectedFixes: rejected,
            hasReverseMotion: hasReverse,
            reason: reason,
            maxDistanceFromRoadM: maxRoadDistance,
            trajectory: trajectory.map { frame.toGeo($0) },
            lastFixSecondsBeforeEnd: lastFixGap
        )
    }

    // MARK: - Transition

    /// The signed-velocity transition, exactly as the baseline runs it:
    ///
    ///     theta <- wrap(theta + (omega - b_omega) dt)
    ///     a      = a_long(theta) - b_a
    ///     ds     = u dt + 0.5 a dt^2
    ///     x     <- x + ds cos(theta)
    ///     y     <- y + ds sin(theta)
    ///     u     <- u + a dt          (signed: reverse is a valid state)
    private func predict(_ state: inout State, control: IMUControl) {
        guard control.dt > 0, !control.gapExceeded else { return }
        let theta = Angles.wrap(state.theta + (control.yawRate - state.gyroBias) * control.dt)
        let a = control.longitudinalAcceleration(heading: theta) - state.accelBias
        let ds = state.u * control.dt + 0.5 * a * control.dt * control.dt
        state.x += ds * cos(theta)
        state.y += ds * sin(theta)
        state.u += a * control.dt
        state.theta = theta
        let growth = config.parkingTracker.processPositionSigmaM * control.dt
        state.positionVariance += growth * growth
    }

    // MARK: - Measurement

    private func reachable(
        fix: TrackerLocation, point: Point,
        accepted: [(fix: TrackerLocation, point: Point)],
        roadPrior: Point, start: Double
    ) -> Bool {
        let cfg = config.parkingTracker
        guard let accuracy = fix.horizontalAccuracy, accuracy > 0, accuracy <= cfg.maxAccuracyM
        else { return false }
        if let previous = accepted.last {
            let dt = Swift.max(0, fix.monotonicTime - previous.fix.monotonicTime)
            let maximum = config.motion.maxSpeedMS * dt + cfg.physicalMarginM
            return point.distance(to: previous.point) <= maximum
        }
        let dt = Swift.max(0, fix.monotonicTime - start)
        let maximum = config.motion.maxSpeedMS * dt + cfg.physicalMarginM
        return point.distance(to: roadPrior) <= maximum
    }

    private func gpsUpdate(_ state: inout State, fix: TrackerLocation, point: Point) {
        let sigma = Swift.max(fix.horizontalAccuracy ?? 8.0, config.gps.minAccuracySigmaM)
        let gain = Swift.min(Swift.max(30.0 / (30.0 + sigma * sigma), 0.08), 0.65)
        let previous = state.point
        state.x += gain * (point.x - state.x)
        state.y += gain * (point.y - state.y)

        if fix.hasValidCourse, fix.hasValidSpeed,
           (fix.speed ?? 0) >= config.parkingTracker.lowSpeedMS {
            state.theta = Angles.headingFromCourse(fix.course ?? 0)
        }
        if fix.hasValidSpeed {
            // CLLocation.speed is an unsigned magnitude. The sign has to come
            // from somewhere else: the direction the tracker actually moved,
            // compared with where it believes it is pointing. Without this a
            // reversing car is integrated forwards.
            let displacement = point - previous
            let direction = Point(x: cos(state.theta), y: sin(state.theta))
            let dot = displacement.x * direction.x + displacement.y * direction.y
            state.u = (dot < 0 ? -1.0 : 1.0) * (fix.speed ?? 0)
        }
    }

    // MARK: - Terminal cluster

    private struct Cluster {
        var points: [Point]
        var weights: [Double]
    }

    /// The set of slow, mutually consistent fixes right at the end of the trip.
    ///
    /// This is what turns "somewhere along the last manoeuvre" into a position:
    /// several low-speed fixes agreeing with each other is strong evidence, and
    /// their scatter is an honest radius.
    private func terminalCluster(
        accepted: [(fix: TrackerLocation, point: Point)], endedAt: Double
    ) -> Cluster? {
        let cfg = config.parkingTracker
        let tail = accepted.filter {
            $0.fix.monotonicTime >= endedAt - cfg.terminalClusterWindowSeconds
                && $0.fix.hasValidSpeed
                && ($0.fix.speed ?? 0) <= cfg.lowSpeedMS
        }
        guard tail.count >= cfg.minClusterFixes else { return nil }

        let medianX = median(tail.map(\.point.x))
        let medianY = median(tail.map(\.point.y))
        let centre = Point(x: medianX, y: medianY)
        let kept = tail.filter { $0.point.distance(to: centre) <= cfg.clusterMedianRadiusM }
        guard kept.count >= cfg.minClusterFixes else { return nil }

        return Cluster(
            points: kept.map(\.point),
            weights: kept.map { 1.0 / pow(Swift.max($0.fix.horizontalAccuracy ?? 5.0, 5.0), 2) }
        )
    }

    private func median(_ values: [Double]) -> Double {
        guard !values.isEmpty else { return 0 }
        let sorted = values.sorted()
        let middle = sorted.count / 2
        // Component-wise median, matching numpy's even-length averaging.
        return sorted.count % 2 == 1
            ? sorted[middle]
            : 0.5 * (sorted[middle - 1] + sorted[middle])
    }

    private func circle(around centre: Point, radius: Double, segments: Int = 48) -> [Coordinate] {
        (0...segments).map { step in
            let angle = 2 * Double.pi * Double(step) / Double(segments)
            return frame.toGeo(
                Point(x: centre.x + radius * cos(angle), y: centre.y + radius * sin(angle))
            )
        }
    }
}
